lib/clerk/authenticate_context.rb

```diff
@@ -0,0 +1,186 @@
+        def frontend_api
+            return "" if !valid_publishable_key?(publishable_key.to_s)
+
+            @frontend_api ||= if !proxy_url.empty?
+                proxy_url
+            elsif development_instance? && !domain.empty?
+                "clerk.#{domain}"
+            else
+                # remove $ postfix
+                decode_publishable_key(publishable_key).chop
+            end
+        end
+
+        def development_instance?
+            # TODO(dimkl): Consolidate with @clerk/backend to use pk instead of sk
+            secret_key.start_with?("sk_test_")
+        end
+
+        def production_instance?
+            # TODO(dimkl): Consolidate with @clerk/backend to use pk instead of sk
+            secret_key.start_with?("sk_live_")
+        end
+
+        def document_request?
+            @headers.sec_fetch_dest == "document"
+        end
+
+        def accepts_html?
+            @headers.accept && @headers.accept.start_with?('text/html')
+        end
+        
+        def eligible_for_multi_domain?
+            is_satellite? && document_request? && !clerk_synced?
+        end
+
+        def active_client?
+            @cookies.client_uat.to_i > 0
+        end
+
+        def cross_origin_request?
+            # origin contains scheme+host and optionally port (omitted if 80 or 443)
+            # ref. https://www.rfc-editor.org/rfc/rfc6454#section-6.1
+            return false if @headers.origin.nil?
+      
+            # strip scheme
+            origin = @headers.origin.strip.sub(/\A(\w+:)?\/\//, '')
+            return false if origin.empty?
+      
+            # Rack's host and port helpers are reverse-proxy-aware; that
+            # is, they prefer the de-facto X-Forwarded-* headers if they're set
+            request_host = @headers.host
+            request_host << ":#{@headers.port}" if @headers.port != 80 && @headers.port != 443
+      
+            origin != request_host
+        end
+
+        def dev_browser?
+            !dev_browser.empty?
+        end
+
+        def session_token_in_header?
+            !session_token_in_header.to_s.empty?
+        end
+
+        def handshake_token?
+            !handshake_token.to_s.empty?
+        end
+
+        def session_token_in_cookie?
+            !session_token_in_cookie.to_s.empty?
+        end
+
+        private
+
+        def valid_publishable_key?(pk)
+            valid_publishable_key_prefix?(pk) && valid_publishable_key_postfix?(pk)
+        end
+
+        def valid_publishable_key_prefix?(pk)
+            pk.start_with?("pk_live_") || pk.start_with?("pk_test_")
+        end
+
+        def valid_publishable_key_postfix?(pk)
+            decode_publishable_key(pk).end_with?("$")
+        end
+
+        def decode_publishable_key(pk)
+            Base64.decode64(pk.split("_")[2].to_s)
+        end
+
+        def retrieve_from_query_string(url, key)
+            request_qs = Rack::Utils.parse_query(url.query)
+            request_qs[key]
```

Comment:
🙃 would probably put this in a single line, without the intermediate variable

Reply:
i will refactor it in the PR with the minor changes / fixes.